### DISCUSSION TITLE
Evaluating efforts for turning on run-time detection for algorithms

### DISCUSSION THREAD

**Discussion by sahanaprasad07** [open-quantum-safe/liboqs#1481] [Q&A]
We know, liboqs provides various compile-time options for the algorithms it implements.

Compile-time detection of platform capabilities gives significant slow-down because of using the sub optimal implementations of the algorithms. OpenSSL uses runtime detection of platform capabilities that ensures better performance.

Could you kindly help me in investigating the options and evaluate efforts for turning it in run-time detection?

Thank you!

**Comment by baentsch** [comment]
Thanks for using `liboqs`. I'm not sure, I understand your questions completely, so allow me to ask back:

> We know, liboqs provides various compile-time options for the algorithms it implements.

That is correct -- but could you please specify which options you're particularly concerned with/use?

> Compile-time detection of platform capabilities gives significant slow-down because of using the sub optimal implementations of the algorithms. 

There is a slow-down due to the runtime detection facility (https://github.com/open-quantum-safe/liboqs/issues/624), yes, but I do not understand what you mean by "using the sub optimal implementations of the algorithms": The facility should use the best available implementation. Can you explain? And maybe also quantify how much slowdown you see -- and under which conditions?

> OpenSSL uses runtime detection of platform capabilities that ensures better performance.

Again: Which features are you referring-to?

> Could you kindly help me in investigating the options and evaluate efforts for turning it in run-time detection?

I do not understand this sentence -- could you please reformulate? 

In general, though, any help improving performance --while retaining portability-- is greatly welcome. We surely can re-open #624 any time if you have suggestions or even PRs for this.

**Comment by beldmit** [comment]
So does, e.g. AVX2 flag means that this processor instructions will be built but not used if runtime detection shows their absence?

**Reply by baentsch** [reply]
Correct. 

All documentation how to control (also to turn off this behaviour and build a YOLO binary, i.e., one that will crash if running on a non-AVX2 platform -- but that is faster and smaller) is [here](https://github.com/open-quantum-safe/liboqs/blob/main/CONFIGURE.md#oqs_dist_build).

**Reply by beldmit** [reply]
`OQS_DIST_BUILD` is the key flag, right?

**Reply by baentsch** [reply]
Yes (hence the link above :). Switch it off and you get the fastest ("YOLO") code for the platform you build on (or the one explicitly configured by OQS_OPT_TARGET). Switch it on and you get code that _always_ runs (with CPU feature detection overhead/penalty). Default is ON.

**Comment by sahanaprasad07** [comment]
Sorry about the fumbling in my question. I got all the answers to my questions through[ this](https://github.com/open-quantum-safe/liboqs/issues/624) Thank you for the quick response!
